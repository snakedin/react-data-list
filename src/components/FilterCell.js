import React, { Component } from 'react';
import PropTypes from 'prop-types';

import DateField from "./filters/DateField";
import ReactSelectField from "./filters/ReactSelectField";
import SelectField from "./filters/SelectField";
import TextField from "./filters/TextField";

export default class FilterCell extends Component {

    static propTypes = {
        filter: PropTypes.oneOfType([
            PropTypes.bool,
            PropTypes.object
        ]),
        value: PropTypes.oneOfType([
            PropTypes.bool,
            PropTypes.string,
            PropTypes.number,
            PropTypes.object,
            PropTypes.array
        ]),
        onChangeFilter: PropTypes.func
    };

    /**
     * Render filter field
     */
    renderField({ type, field, ...params }) {
        const { value, onChangeFilter } = this.props;

        if (type === 'custom') {
            const { handler } = params;
            return handler(field, value, onChangeFilter);
        } else {
            let Element;

            switch (type) {
                case 'select':
                    Element = SelectField;
                    break;

                case 'react-select':
                    Element = ReactSelectField;
                    break;

                case 'date':
                    Element = DateField;
                    break;

                case 'text':
                default:
                    Element = TextField
            }

            return <Element id={ field } value={ value } onChange={ onChangeFilter } { ...params }/>
        }
    }

    /**
     * Render cell content
     */
    render() {

        const { filter } = this.props;

        return  (
            <td>
                { this.renderField(filter) }
            </td>
        );
    }
}