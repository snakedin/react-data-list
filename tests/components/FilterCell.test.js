import React from "react";
import { render } from "@testing-library/react";
import FilterCell from "../../src/components/FilterCell";

it("FilterCell render", () => {

    function tdRender(component) {
        const tableRow = document.createElement('tr');

        const { container } = render(component, {
            container: document.body.appendChild(tableRow)
        });

        return container;
    }

    const textField = tdRender(<FilterCell filter={{type: 'text', field: 'name'}}/>);
    expect(textField).toMatchSnapshot();

    const selectFilter = {
        type: 'select',
        field: 'name',
        items: [
            {value: 1, label: 'Item 1'},
            {value: 2, label: 'Item 2'},
            {value: 3, label: 'Item 3'}
        ]
    };
    const selectField = tdRender(<FilterCell value={'2'} filter={selectFilter}/>);
    expect(selectField).toMatchSnapshot();
});