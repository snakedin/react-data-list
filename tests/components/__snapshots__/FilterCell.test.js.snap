// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterCell render 1`] = `
<tr>
  <td>
    <input
      class="form-control"
      type="text"
      value=""
    />
  </td>
</tr>
`;

exports[`FilterCell render 2`] = `
<tr>
  <td>
    <select
      class="form-control"
    >
      <option
        value="1"
      >
        Item 1
      </option>
      <option
        selected=""
        value="2"
      >
        Item 2
      </option>
      <option
        value="3"
      >
        Item 3
      </option>
    </select>
  </td>
</tr>
`;
